src/Transport/Sending/MessageDispatcher.cs

```diff
@@ -66,6 +71,36 @@ public Task Dispatch(TransportOperations outgoingMessages, TransportTransaction
             return tasks.Count == 1 ? tasks[0] : Task.WhenAll(tasks);
         }
 
+        static void AssertBelowMaxMessageThresholdForTransaction(List<UnicastTransportOperation> unicastTransportOperations, List<MulticastTransportOperation> multicastTransportOperations, CommittableTransaction committableTransaction)
+        {
+            var totalNumberOfOperations = unicastTransportOperations.Count + multicastTransportOperations.Count;
+            if (committableTransaction == null || totalNumberOfOperations <= MaxMessageThresholdForTransaction)
+            {
+                return;
+            }
+
+            var numberOfTransactionalOperations = 0;
+            foreach (var transportOperation in unicastTransportOperations)
```

Comment:
```suggestion
             // we are iterating through the operations twice, the tradeoff here is that we rather enumerate twice
             // than incurring IO costs in scenarios we know it will fail
            var numberOfTransactionalOperations = 0;
            foreach (var transportOperation in unicastTransportOperations)
```

Reply:
@danielmarbach Since this is comment refers to the other alternatives, we prefer to document this in an ADR.